common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java

```diff
@@ -40,24 +40,41 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
     }
   }
 
+  private static final boolean unaligned = Platform.unaligned();
   /**
    * Optimized byte array equality check for byte arrays.
    * @return true if the arrays are equal, false otherwise
    */
   public static boolean arrayEquals(
       Object leftBase, long leftOffset, Object rightBase, long rightOffset, final long length) {
     int i = 0;
-    while (i <= length - 8) {
-      if (Platform.getLong(leftBase, leftOffset + i) !=
-        Platform.getLong(rightBase, rightOffset + i)) {
-        return false;
+
+    // check if stars align and we can get both offsets to be aligned
+    if ((leftOffset % 8) == (rightOffset % 8)) {
```

Comment:
@srowen sorry for the so late comment, shall we add `!unaligned` here to fast-fail the condition ?

Reply:
This is the case where it's already aligned, so works either way?

Reply:
I just consider it can avoid run `(leftOffset % 8) == (rightOffset % 8)` and even `while ((leftOffset + i) % 8 != 0 && i < length) {`  if we can break the branch fast since most Platform is unaligned.

Reply:
I think the idea was to avoid unaligned access even where it is possible, as it used to be faster. I'm not sure it matters much now. Maybe more on ARM. If there's evidence it's actually faster to skip this and not align, OK. Otherwise without being sure I'd leave it at trying to align

Reply:
I created https://github.com/apache/spark/pull/35078 for the further perf number